plugins/gradle/resources/org/jetbrains/plugins/gradle/GradleImprovedHotswapDetectionInitScriptUsingService.groovy

```diff
@@ -0,0 +1,163 @@
+
+  void registerTask(Task task) {
+    tasks.add(task)
+    if (task.outputs.hasOutput) {
+      beforeTaskFileHashes.put(task, hashOutputs(task.outputs.files.files))
+    }
+  }
+
+  // this runs after *all* tasks have executed
+  @Override
+  void close() throws Exception {
+    def outputFile = getParameters().outputFile.get()
+    tasks.each { Task task ->
+      def state = task.state
+      def work = state.didWork
+      def fromCache = state.skipped && state.skipMessage == 'FROM-CACHE'
+      def hasOutput = task.outputs.hasOutput
+      if ((work || fromCache) && hasOutput) {
+        def beforeHashes = beforeTaskFileHashes[task]
+        task.outputs.files.files.each { File outputDir ->
+          if (outputDir.directory) {
+            def hashes = hashOutputDir(outputDir)
+            hashes.each { file, hash ->
+              def beforeHash = beforeHashes[file]
+              if (beforeHash == null || beforeHash != hash) {
+                def relativePath = outputDir.toPath().relativize(file.toPath())
+                outputFile.append("root:${outputDir.path}\n")
+                outputFile.append("path:${relativePath}\n")
+              }
+            }
+          } else {
+            def hash = hashFile(outputDir)
+            def beforeHash = beforeHashes[outputDir]
+            if (beforeHash == null || beforeHash != hash) {
+              outputFile.append("root:${outputDir.path}\n")
+              outputFile.append("path:\n")
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private static Map<File,String> hashOutputs(Set<File> outputs) {
+    Map<File,String> hashes = [:]
+
+    outputs.each { outputDir ->
+      if (outputDir.isDirectory()) {
+        hashFilesInDir(outputDir, hashes)
+      }
+    }
+    return hashes
+  }
+
+  private static Map<File,String> hashOutputDir(File outputDir) {
+    Map<File,String> hashes = [:]
+    hashFilesInDir(outputDir, hashes)
+    return hashes
+  }
+
+  private static hashFilesInDir(File dir, Map<File,String> hashes) {
+    dir.eachFileRecurse(FILES) {
+      hashes.put(it, hashFile(it))
+    }
+  }
+
+  private static String hashFile(File file) {
+    // Set your algorithm
+    // "MD2","MD5","SHA","SHA-1","SHA-256","SHA-384","SHA-512"
+    MessageDigest hash = MessageDigest.getInstance("SHA-256")
+
+    FileInputStream fis = new FileInputStream(file)
+    try {
+      byte[] fileBuffer = new byte[1024]
+
+      int bytesRead
+      while ((bytesRead = fis.read(fileBuffer)) != -1) {
+        hash.update(fileBuffer, 0, bytesRead)
+      }
+
+      return hash.digest().encodeHex()
+    } finally {
+      fis.close()
+    }
+  }
+}
+
+Provider<OutputPathCollectorService> provider = gradle.sharedServices.registerIfAbsent("outputPathCollectorService",
+                                                                                       OutputPathCollectorService) { it.parameters.outputFile.set(outputFile)  }
+
+gradle.taskGraph.whenReady { TaskExecutionGraph tg ->
```

Comment:
this whole thing is copied because of this section? weird x/ almost seems like you should just concat em

Reply:
_ideally_ this whole init script should be using compiled code instead of dumping everything into the init script.

However it's not straightforward to load random code into init scripts

I think since there are only 2 files I wont' try to optimize it just yet

Reply:
yep that's fair